### PR TITLE
Pulsar

### PR DESCRIPTION
**Pull request by ric-evans**
Two major developments
1. Added pulsar backend (with unit tests)
2. Added integration tests for all backends

### CODE REVIEW

File: MQClient/backends/pulsar.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        except Exception as e:
+            if str(e) == "Pulsar error: TimeOut":  # pulsar isn't a fan of derived Exceptions
+                return None
+            if str(e) == "Pulsar error: AlreadyClosed":
+                queue.connect()
+                continue
+            raise
+
+        break  # poor man's do-while loop
```
Comment (blinkdog) [inline MQClient/backends/pulsar.py:136]: Is the `break` at the end reachable?
It seems like the try block would return a Message object, return None, or exception out.
It seems like the except block would return None, jump back up to while True, or exception out by raising the original exception.

File: MQClient/backends/pulsar.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+            try:
+                yield msg
+            except Exception as e:  # pylint: disable=W0703
+                reject_message(queue, msg.msg_id)
+                if propagate_error:
+                    raise
+                logging.warning('error downstream: %r', e, exc_info=True)
+            else:
+                if auto_ack:
+                    ack_message(queue, msg.msg_id)
```
Comment (blinkdog) [inline MQClient/backends/pulsar.py:187]: This seems a little odd; can `yield msg` actually raise an Exception?

Comment (dsschult) [reply MQClient/backends/pulsar.py:187]: This is a generator, so the `yield msg` is passing control to somewhere else, which may raise an Exception.

Comment (dsschult) [reply MQClient/backends/pulsar.py:187]: I think we're missing a unit test that triggers this though.

Comment (dsschult) [reply MQClient/backends/pulsar.py:187]: Trying to make a code snippet that does this, I may have gotten generators and context managers mixed together.  It doesn't seen like generators can catch exceptions.

So, we probably need to change how this function works.  But I recommend another pull request for that.

File: MQClient/backends/pulsar.py (hunk 1 above)
Comment (ric-evans) [reply MQClient/backends/pulsar.py]: Good catch. I'm surprised one off my linters wasn't able to catch this.

File: MQClient/backends/pulsar.py (hunk 2 above)
Comment (ric-evans) [reply MQClient/backends/pulsar.py:188]: This function was essentially lifted straight from RabbitMQ. So, it makes sense to punt this

**Pull request merged**